### PR TITLE
[1.20.5] Implement specialized DataComponentType DeferredRegister

### PR DESCRIPTION
**Pull request by dhyces**
This PR aims to add a new specialized `DeferredRegister` for `DataComponentType`s. Due to data component types having a generic type, they require explicit type parameters when using regular registration methods. This PR fixes this using vanilla's own solution, a `UnaryOperator` with proper type inference. The `DeferredDataComponentType` also implements `DataComponentType` so that users may simply pass the deferred object to item property methods without calling `get`, which is simple convenience.

### CODE REVIEW

**Comment by PepperCode1** [comment]
I don't think implementing the registered object type (`DataComponentType`) on the holder is a good idea because the holders are not themselves registered so they cannot be used to look up the `ResourceLocation` of the registered object. The expectation when any code has access to an instance of any registered object type (such as `DataComponentType`) is that it has an associated `ResourceLocation` in the registry.

**Comment by dhyces** [comment]
> I don't think implementing the registered object type (`DataComponentType`) on the holder is a good idea because the holders are not themselves registered so they cannot be used to look up the `ResourceLocation` of the registered object. The expectation when any code has access to an instance of any registered object type (such as `DataComponentType`) is that it has an associated `ResourceLocation` in the registry.

That's a good point, forgot about that when considering whether it was safe to do so. Removing it now.

EDIT: Actually, looking at it now, is there any reason to have a convenience type for this if it has no additional usable methods? Sure, it cuts down on generic verbosity, but otherwise it's just a regular `DeferredHolder`. Might as well just be a specialized registration method?

**Comment by ApexModder** [comment]
> EDIT: Actually, looking at it now, is there any reason to have a convenience type for this if it has no additional usable methods? Sure, it cuts down on generic verbosity, but otherwise it's just a regular DeferredHolder. Might as well just be a specialized registration method?

Yeah a specialized `DeferredHolder` type doesnt make much sense to me, the specialized register method is good though.
Could maybe also add a few other register methods which take a `Codec`/`StreamCodec` and wrap that into `UnaryOperator` method? 

<details>

<summary>Examples</summary>

```java
public <T> DeferredHolder<DataComponentType<?>, DataComponentType<T>> register(String registrationName, UnaryOperator<DataComponentType.Builder<T>> builder) {
    return register(registrationName, () -> builder.apply(DataComponentType.builder()).build());
}

public <T> DeferredHolder<DataComponentType<?>, DataComponentType<T>> register(String registrationName, Codec<T> codec, StreamCodec<? super RegistryFriendlyByteBuf, T> networkCodec) {
    return register(registrationName, builder -> builder.persistent(codec).networkSynchronized(networkCodec));
}

public <T> DeferredHolder<DataComponentType<?>, DataComponentType<T>> register(String registrationName, Codec<T> codec, Codec<T> networkCodec) {
    return register(registrationName, codec, ByteBufCodecs.fromCodecWithRegistries(networkCodec));
}

public <T> DeferredHolder<DataComponentType<?>, DataComponentType<T>> register(String registrationName, Codec<T> codec) {
    return register(registrationName, codec, codec);
}
```

</details>

To allow passing in the deferred type could patch in `Supplier` based component type get/set/has methods much like attachment types have, that way it would look up the correct registered object via the `Supplier` implementation, rather than using the deferred object itself.


Edit: Also looks like you have license header issues, run `gradlew applyAllFormatting` to apply neos spotless config, re-formats the code and applies licensing headers where need be, will get an error if theres any formatting violations.

and `gradlew unpackSourcePatches` to generate patch files
[_this is only needed if you modified vanilla code in `./projects/neoforge`, and best to be run before applying spotless config_]

File: src/main/java/net/neoforged/neoforge/registries/DeferredRegister.java [hunk 1]
```diff
@@ -580,6 +595,32 @@ protected <I extends Item> DeferredItem<I> createHolder(ResourceKey<? extends Re
...
+        protected DataComponents(String namespace) {
+            super(Registries.DATA_COMPONENT_TYPE, namespace);
+        }
+
+        /**
+         * Convenience method that constructs a builder for use in the operator. Use this to avoid inference issues.
+         * 
+         * @param name    The name for this data component type. It will automatically have the {@linkplain #getNamespace() namespace} prefixed.
+         * @param builder The unary operator, which is passed a new builder for user operations, then builds it when registered.
+         * @return A {@link DeferredDataComponentType} that will track updates from the registry for this type.
```
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/registries/DeferredRegister.java]: there are no "registry updates" anymore

File: src/main/java/net/neoforged/neoforge/registries/DeferredRegister.java [hunk 2]
```diff
@@ -580,6 +595,32 @@ protected <I extends Item> DeferredItem<I> createHolder(ResourceKey<? extends Re
...
+         */
+        public <D> DeferredDataComponentType<D> registerBuilder(String name, UnaryOperator<DataComponentType.Builder<D>> builder) {
```
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/registries/DeferredRegister.java]: the naming is not great - it's an actual component type that will be registered, not a builder

**Review by Technici4n — CHANGES_REQUESTED** [review]
I'm not sure that this feature is worth having as it makes the DeferredRegister API somewhat longer. It's probably fine given how widely used I'd expect data components to be.

File: src/main/java/net/neoforged/neoforge/registries/DeferredRegister.java (hunk 1 above)
Comment (dhyces) [reply src/main/java/net/neoforged/neoforge/registries/DeferredRegister.java]: Docs for this were adapted from Items

File: src/main/java/net/neoforged/neoforge/registries/DeferredRegister.java (hunk 2 above)
Comment (dhyces) [reply src/main/java/net/neoforged/neoforge/registries/DeferredRegister.java]: The object that the user will ever interact with in this case is a builder. Hence the shortened name `registerBuilder` as opposed to something like `registerWithBuilder`. Though, if I think of another name or another name is suggested, I can update it

**Comment by dhyces** [comment]
> > EDIT: Actually, looking at it now, is there any reason to have a convenience type for this if it has no additional usable methods? Sure, it cuts down on generic verbosity, but otherwise it's just a regular DeferredHolder. Might as well just be a specialized registration method?
> 
> Yeah a specialized `DeferredHolder` type doesnt make much sense to me, the specialized register method is good though.
> Could maybe also add a few other register methods which take a `Codec`/`StreamCodec` and wrap that into `UnaryOperator` method? 
> 
> <details>
> 
> <summary>Examples</summary>
> 
> ```java
> public <T> DeferredHolder<DataComponentType<?>, DataComponentType<T>> register(String registrationName, UnaryOperator<DataComponentType.Builder<T>> builder) {
>     return register(registrationName, () -> builder.apply(DataComponentType.builder()).build());
> }
> 
> public <T> DeferredHolder<DataComponentType<?>, DataComponentType<T>> register(String registrationName, Codec<T> codec, StreamCodec<? super RegistryFriendlyByteBuf, T> networkCodec) {
>     return register(registrationName, builder -> builder.persistent(codec).networkSynchronized(networkCodec));
> }
> 
> public <T> DeferredHolder<DataComponentType<?>, DataComponentType<T>> register(String registrationName, Codec<T> codec, Codec<T> networkCodec) {
>     return register(registrationName, codec, ByteBufCodecs.fromCodecWithRegistries(networkCodec));
> }
> 
> public <T> DeferredHolder<DataComponentType<?>, DataComponentType<T>> register(String registrationName, Codec<T> codec) {
>     return register(registrationName, codec, codec);
> }
> ```
> 
> </details>
> 
> To allow passing in the deferred type could patch in `Supplier` based component type get/set/has methods much like attachment types have, that way it would look up the correct registered object via the `Supplier` implementation, rather than using the deferred object itself.
> 
> 
> Edit: Also looks like you have license header issues, run `gradlew applyAllFormatting` to apply neos spotless config, re-formats the code and applies licensing headers where need be, will get an error if theres any formatting violations.
> 
> and `gradlew unpackSourcePatches` to generate patch files
> [_this is only needed if you modified vanilla code in `./projects/neoforge`, and best to be run before applying spotless config_]

Thanks for a reminder about the licensing, I had forgotten that part of the workflow. I know how to apply patches and this PR was not in need of any patch updates.

**Comment by dhyces** [comment]
> I'm not sure that this feature is worth having as it makes the DeferredRegister API somewhat longer. It's probably fine given how widely used I'd expect data components to be.

I'm inclined to agree, which is why I had mentioned a simple registration method instead. Only problem with that solution is that it bloats the regular generic registration methods with a specialized method which could only be used in registering data component types

File: src/main/java/net/neoforged/neoforge/registries/DeferredRegister.java (hunk 2 above)
Comment (Technici4n) [reply src/main/java/net/neoforged/neoforge/registries/DeferredRegister.java]: could be `registerComponent`

Comment (ApexModder) [reply src/main/java/net/neoforged/neoforge/registries/DeferredRegister.java]: why not simply call it `register`, its already inside of a type specific register `DeferredRegister.DataComponents`.

`Blocks` and `Items` dont call theirs `registerItem` or `registerBlock` so why should this be any different?
and the places where they do, are only named differently cause parameters clash or its for a sub-type (item block)

Comment (Technici4n) [reply src/main/java/net/neoforged/neoforge/registries/DeferredRegister.java]: image

They do call them like that... The ones that are called `register` are covariant overrides from `DeferredRegister`.

Comment (dhyces) [reply src/main/java/net/neoforged/neoforge/registries/DeferredRegister.java]: > why not simply call it `register`, its already inside of a type specific register `DeferredRegister.DataComponents`.
> 
> `Blocks` and `Items` dont call theirs `registerItem` or `registerBlock` so why should this be any different? and the places where they do, are only named differently cause parameters clash or its for a sub-type (item block)

In this case, it does clash, due to type erasure from `Function<RL, T>` and `UnaryFunction<T>` which extends `Function<T,T>`

File: src/main/java/net/neoforged/neoforge/registries/DeferredRegister.java [hunk 3]
```diff
@@ -580,6 +595,32 @@ protected <I extends Item> DeferredItem<I> createHolder(ResourceKey<? extends Re
...
+        }
+
+        /**
+         * Convenience method that constructs a builder for use in the operator. Use this to avoid inference issues.
+         * 
+         * @param name    The name for this data component type. It will automatically have the {@linkplain #getNamespace() namespace} prefixed.
+         * @param builder The unary operator, which is passed a new builder for user operations, then builds it upon registration.
+         * @return A {@link DeferredDataComponentType} which reflects the data that will be registered.
+         */
+        public <D> DeferredDataComponentType<D> registerComponent(String name, UnaryOperator<DataComponentType.Builder<D>> builder) {
```
Comment (Minecraftschurli) [inline src/main/java/net/neoforged/neoforge/registries/DeferredRegister.java]: rename to `registerComponentType`

**Review by Minecraftschurli — CHANGES_REQUESTED** [review]

**Comment by Minecraftschurli** [comment]
idk if the dedicated deferred holder is needed

**Review by Minecraftschurli — APPROVED** [review]

**Review by Technici4n — APPROVED** [review]

**Pull request merged**